### ISSUE TITLE
Scheduled tasks should avoid overloading the streamprocessor

### ISSUE THREAD

**Issue by korthout** [camunda/zeebe#13870]
**Description**

The workflow engine employs the help of several scheduled tasks that periodically append commands to the log stream. This helps offload time-consuming work that could otherwise block the actual workflow processing. 

For example, job deadlines are checked for expiration by the `JobTimeoutTrigger`, which will append commands for each job to time them out. The stream processor will then process this command asynchronously. Eventually, the job changes state to `TIMED_OUT` when the command is processed. 

In the meantime, the `JobTimeoutTrigger` continues to check for expired job deadlines periodically. If the stream processor is blocked, it may happen that the `JobTimeoutTrigger` continues to append the same `Job:TIME_OUT` command. There is nothing that stops it from overloading the stream processor.

In fact, many of the scheduled tasks suffer from a similar problem.

**Proposed solution**

Scheduled tasks need to keep a cache of recently appended commands (or something similar) that can be used to avoid appending the same command multiple times.

This should be available in a generic way, so all scheduled tasks can easily add support for this.

**Incident**

Follow up issue from incident: [INC-274: Streamprocessor lagging behind](https://docs.google.com/document/d/1l8CPBThWtPUxEpmwcoxeb_9pn1eRB2CPpHPiwEVoa3Q/edit#heading=h.hyf892uforsw)

Support: 
- https://jira.camunda.com/browse/SUPPORT-18291
- https://jira.camunda.com/browse/SUPPORT-18997

**Comment by korthout** [comment]
Marking this as high impact, as it is a recurring pain. Resolving this would lower the load on clusters experiencing problems where timers, buffered messages, or job timeouts play a role. This would help in the investigation as well as ease recovery.

**Comment by abbasadel** [comment]
I'm increasing the priority of this issue/feature to 'upcoming' since there is a support case waiting for it. 
Still, it is highly unlikely to be able to work on this before 8.4. However, we will update the issue once we plan to work on it.

**Comment by megglos** [comment]
ZDP-Planning:
- pushing into the inbox of ZPA to get an update on this as it relates to #14003

**Comment by korthout** [comment]
The ZPA-team discussed this issue again and now considers this a bug. Originally, I felt it was a feature, as we have never supported many timers triggering at once (i.e., there's always a limit somewhere). However, the team feels that outages should not occur from existing features, irrespective of how they are used. Add to this that we don't have documentation about some artibrary limit to the number of timers. Lastly, users feel that this is a bug, and so should we.

As a bug, it may be easier to prioritize and release a fix for this as we can backport and patch previous versions.

**Comment by dddpaul** [comment]
@korthout Hi!

Is there any updates for this bug? I would be nice to have a due date )

**Comment by korthout** [comment]
Hi @dddpaul, the team plans to spend time on this in our current iteration. There's a chance that we can fit a solution in the upcoming patch releases, but we cannot promise this.

We'll need to consider different solutions. For example, the proposed cache reduces the load on the system in most cases but will not remove the problem completely. Alternatively, ideas like #12560 could help but come with other downsides. The developer(s) working on this issue should consider whether caching is enough or whether another solution might be more fitting. As this is not yet clear, it's possible that a solution will not be ready before the upcoming patch releases. However, we plan to spend time on this issue now.

**Comment by Zelldon** [comment]
There was an L1 blocker again because of this https://jira.camunda.com/plugins/servlet/mobile#issue/SUPPORT-18997 

Handled by @npepinpe

**Comment by npepinpe** [comment]
Users are heavily impacted by this, so I would like to bump this to critical, with a focus on fixing timers. Timers are different from message correlation has they don't involve inter-partition communication, which may help narrow the scope of the fix for now (i.e. writes are guaranteed to succeed).

That said, I noticed some possible concurrency issues with the `DueDateTimerChecker`, which could lead to more timers than required being scheduled under high load.

If enabled, the `DueDateTimerChecker` runs in the asynchronous scheduler actor, separately from the stream processor.

However, both actors will access the `DueDateTimerChecker`:

`DueDateTimerChecker#scheduleTimer` is called from the engine/side-effects, i.e. processing actor. It then:

- Accesses the following state: `shouldRescheduleChecker`, `checkerRunning`, `nextDueDate` (rw)
- Checks if the checker is currently running. This is a non-volatile boolean check.
- If it's not running, it schedules a task.

At the same time, scheduling is done on the async scheduler actor. This is the `TriggerEntitiesTask`. It:

- Accesses the following state: `shouldRescheduleChecker`, `checkerRunning` (rw), `nextDueDate` (rw)

Is this correct? If so we have non-volatile state being accessed by two different actors, and possibly some race conditions on when and if to schedule the next timer.

Moving on the the possible solution. In order to keep it light of course, we'd need to to have a way to identify a timer trigger uniquely. Since every time instance can only be scheduled for trigger once in the state, it should be enough to distinguish them with the timer key, which is a single long. 

The general idea, then, is to cache locally in the `DueDateTimerChecker` state (or some state) which commands have been written, so as to avoid writing them again. A perfect cache would be unbounded - you'd have to possibly store as many timers in there as could possibly exist (worst case scenario).  But a lossy cache is probably acceptable. Take `LongHashSet` from Agrona - an empty, pre-initialized set instance of 1 million keys takes 8MB of memory. Filled, it takes about twice, ~16MB. It's a bit hard to add LRU behavior to such a set however. Eviction could be random to keep things simple/fast, or we can use a `SortedSet`. However, a `TreeSet` takes about ~60MB for 1 million longs, so it's already substantially heavier - but it's pretty easy/fast to add LRU eviction to it. So on the off chance the key was evicted _before_ it was removed explicitly from the cache, you'd potentially still write a second trigger command. But the bigger the cache, the less likely this is to happen.

I sync'd with @Zelldon today about this, and we decided to move forward with the local cache, only for checkers which perform no inter-partition communication.

For timers, we'll use the timer key as the unique cache ID. When iterating of timers, if the timer key is in the cache, then we skip it. Once the task result is built, we can populate the cache with all the timer keys we will be writing. Since the scheduler (even async) is sequenced, we're guaranteed that this write will succeed, or the whole thing will fail (and the cache would be "restarted"). We won't be touching things like leader election and so on. On recovery, it's still possible that the same timers would be triggered twice, but the overload load will be greatly reduced.

This solution is likely adaptable to other checkers which run locally, such as job timeout and message expiry. For messages we can easily use the key. For jobs, we can't use the job key, since you can have multiple activations/time outs. We could use the job key if we can guarantee that the cache is properly cleaned up whenever the job is activated.

Anyway, long winded comment to say we'll get started on a bounded lossy cache-based prototype.

**Comment by npepinpe** [comment]
So in the end, we'll build the cache on the stream processor side. I describe the approach we ended up with in the description of this PR: https://github.com/camunda/zeebe/pull/15073

Let me know if you have questions. For now it's entirely on the stream platform/ZDP side, but it's still useful to know how it works. Plus, if you want to cache future scheduled commands in the future, you'll need to know where to add them :upside_down_face:

**Comment by npepinpe** [comment]
See some benchmark results courtesy of @Zelldon - https://github.com/camunda/zeebe/pull/15073#issuecomment-1803887603 :tada:

**Comment by korthout** [comment]
We took it off the ZPA board, as ZDP has found a solution on their side 🙇  Thanks again @Zelldon and @npepinpe

**Comment by dddpaul** [comment]
Hi!

This looks like a wonderful job.  But this new timer cache doesn't solve my problem with bunch of UNIQUE timers from DIFFERENT process instances. I've illustrated my problem here https://github.com/camunda/zeebe/issues/5134#issuecomment-1752006281. My comment was stated as duplicate of this issue which was likely fixed with timer cache solution in PR #15136 (for 8.3).

I could not understand how this cache solution will help me but had some hope and belief. Well, there no miracle happened.
I've just tested on 8.3-stable
![image](https://github.com/camunda/zeebe/assets/628226/19ab7e3f-7708-465b-8bce-215e55fd6278)

Thousands or millions of unique timer events still overload brokers:
![image](https://github.com/camunda/zeebe/assets/628226/97951508-56da-4723-8910-3ee8149461eb)

This is my process with timer:
![image](https://github.com/camunda/zeebe/assets/628226/471fd3e9-c5c5-4c0b-b4c8-e0a9263259b7)

**Comment by npepinpe** [comment]
We're working on another solution for that in parallel. https://github.com/camunda/zeebe/issues/10087

I know the issue specifies IPC, but the proposed solution will enforce rate limits between the different categories, user submitted commands, internal/scheduled commands, and remote ones.
